packages/analytics-js/src/components/userSessionManager/UserSessionManager.ts

```diff
@@ -21,8 +21,128 @@ class UserSessionManager implements IUserSessionManager {
     this.storage = storage;
   }
 
-  setStorage(storage: IStore) {
+  /**
+   * Initialize User session with values from storage
+   * @param storage Selected storage
+   */
+  init(storage: IStore) {
     this.storage = storage;
+
+    // get the values from storage and set it again
+    this.setUserId(this.getUserId() || '');
+    this.setUserTraits(this.getUserTraits() || {});
+    this.setGroupId(this.getGroupId() || '');
+    this.setGroupTraits(this.getGroupTraits() || {});
+    this.setAnonymousId(this.getAnonymousId());
+
+    const initialReferrer = this.getInitialReferrer();
+    const initialReferringDomain = this.getInitialReferringDomain();
+
+    if (initialReferrer && initialReferringDomain) {
+      this.setInitialReferrer(initialReferrer);
+      this.setInitialReferringDomain(initialReferringDomain);
+    } else {
+      if (initialReferrer) {
+        this.setInitialReferrer(initialReferrer);
+        this.setInitialReferringDomain(getReferringDomain(initialReferrer));
+      }
+      const referrer = getReferrer();
+      this.setInitialReferrer(referrer);
+      this.setInitialReferringDomain(getReferringDomain(referrer));
+    }
+    // Register the effect to sync with storage
+    this.syncSessionWithStorage();
+  }
+
+  /**
+   * Function to update storage whenever state value changes
+   */
+  syncSessionWithStorage() {
+    /**
+     * Update userId in storage automatically when userId is updated in state
+     */
+    effect(() => {
+      if (state.session.userId.value) {
+        this.storage?.set(userSessionStorageKeys.userId, state.session.userId.value);
+      } else {
+        this.storage?.remove(userSessionStorageKeys.userId);
+      }
+    });
```

Comment:
```suggestion
    effect(() => {
      if (state.session.userId.value !== '' && state.session.userId.value !== undefined && state.session.userId.value !== null) {
        this.storage?.set(userSessionStorageKeys.userId, state.session.userId.value);
      } else {
        this.storage?.remove(userSessionStorageKeys.userId);
      }
    });
```

Reply:
Let's create a utility for the valid data check to avoid redundant code.

Reply:
Not required anymore